File: 백준/Silver/9012. 괄호/괄호.swift
import Foundation
var n = readLine()

for _ in 1...(Int(n ?? "") ?? 0) {
    var stack: [String] = []
    var isPS: Bool = true
    if let inputPS = readLine() {
        for i in inputPS {
            if i == "(" {
                stack.append(String(i))
            } else {
                if stack.isEmpty == true {
                    isPS = false
                    break
                } else {
                    stack.popLast()
                    isPS = true
                }
                
            }
        }
        isPS == true && stack.isEmpty == true ? print("YES") : print("NO")
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

from collections import deque

for _ in range(int(input())):
    brackets = input().rstrip()
    stack = deque()

    for bracket in brackets:                # 괄호를 하나씩 뽑아,
        if bracket == '(':                  # 여는 괄호라면,
            stack.append(bracket)           # 스택에 산입
        
        elif stack:                         # 닫는 괄호 + 스택이 비지 않았다면,
            stack.pop()                     # 여는 괄호 스택에서 pop

        else:                               # 닫는 괄호인데 스택이 비었다면, VPS 불가
            stack = True                    # stack에 True를 할당하고
            break                           # break

    print("YES" if not stack else "NO")